### ISSUE TITLE
DS18B20 Issue

### ISSUE THREAD

**Issue by mfittipaldi90** [kizniche/Mycodo#1393]
### Describe the problem/bug

DS18B20 Temperature sensor stopped working overnight and got stuck at 0C. Config page for w1thermsensor shows two "00-00000000." device ID
Tried a config page for OW-Shell and device ID shows 2 different unique IDs

The input with ow-shell library provides erratic measurements ranging from 0C to 96C jumping all around.

I bought a new sensor (both from adafruit) and same behavior happens with brand new sensor. Sensors are installed with 4.7k pullup resistor between signal and VCC.

### Versions:

 - Mycodo Version: 8.15.13
 - Raspberry Pi Version: 4b
 - Raspbian OS Version: *[e.g. Buster Lite]*

### Screenshots
![image](https://github.com/user-attachments/assets/ae56c1fb-de98-44fe-a1fd-b95169d4b289)
![image](https://github.com/user-attachments/assets/c3243b71-ad1f-4e71-aaa2-fdf823d80bf7)

### Additional context

Logs show:
ERROR - mycodo.inputs.ds18b20_076d8c59 - Error 101: Device not set up. See https://kizniche.github.io/Mycodo/Error-Codes#error-101 for more info.

**Comment by kizniche** [comment]
Please describe the details of your wiring.

**Comment by mfittipaldi90** [comment]
Sorry for the bad picture and diagram. This setup was working very well for the past weeks and just stopped working overnight. 
![Wiring](https://github.com/user-attachments/assets/9f6feef2-676e-4992-b1e5-ade3ec08f8ef)

**Comment by kizniche** [comment]
What is the length of wire? Have you tried 5 VDC?

**Comment by kizniche** [comment]
Have you verified the resistance of the resistor?

**Comment by mfittipaldi90** [comment]
Yes, tried the 5 VDC and still same behavior
Length of the wire is the length that came with the sensor itself plus the jumper wire that was soldered in.
Tried two different resistors that are 4.7k ea (both verified with multimeter).
Just tried reflashing my rpi and installing a fresh new mycodo and still not working... now I have to reconfigure it all again :)

**Comment by kizniche** [comment]
Have you tried one sensor at a time?

**Comment by mfittipaldi90** [comment]
Yes, right now I have only one DS18B20 connected and still same behavior. Even swapping DS18B20 and behavior is the exact same. Keep getting phantom IDs and no data reads

**Comment by kizniche** [comment]
You've enabled 1wire after reinstalling?

**Comment by mfittipaldi90** [comment]
yes, it is enabled

**Comment by kizniche** [comment]
Sounds like damaged hardware. Have you tried other software and hardware to communicate with the sensors?

**Comment by mfittipaldi90** [comment]
Ouch, guess I will have to find a new RPi. I will try that. Thank you so much for the quick responses! Sent from my iPhoneOn Aug 24, 2024, at 6:45 PM, Kyle Gabriel ***@***.***> wrote:﻿
Sounds like damaged hardware. Have you tried other software and hardware to communicate with the sensors?

**Comment by kizniche** [comment]
I would try assigning 1wire to another pin besides 4 and testing. Do you have an arduino-compatible microcontroller?

**Comment by kizniche** [comment]
And by try other software, there are many available to communicate with 1wire sensors, the ds18b20 in particular, since it's probably the most populate 1wire sensor in existence.

**Comment by kizniche** [comment]
There's also the possibility you have counterfeit sensors (very common). Did you buy from a reputable vendor?

**Comment by mfittipaldi90** [comment]
I bought from Adafruit… pretty reputable. I purchased a new rpi 4 to test
it out as I do not have an arduino or other microcontroller to test. I have
use for the other rpi so it will come in handy.

I tried a different gpio pin but got the same result. Digging through rpi
forums i found a similar issie that was the rpi. I’ll test the new rpi
that’s coming in soon and report back for completion so we can close this
for future reference.

Mauro Fittipaldi, PhD
786-339-2074

**Comment by wbard93** [comment]
I've had various problems with the DS18B20 stopping to work on the RPi. Usually it works for a while and the just stops. At least once it started working again when I disabled the input in Mycodo and re-enabled it but usually it starts working when I reboot the RPi. I remember one instance where using RPi routine independent of Mycodo I was able to read the temp and recovered it by reimaging the SD and reloading Mycodo.

I think one time that it stopped working, the RPi was bad. I've also had broken connections that have caused it to stop working.

I suspect at least part of the problem is a bug in the RPI DS18B20 library I'm using. In my view, it's not very reliable and is likely a result of the multi tasking RPi trying to talk using a bit-bang serial interface. My thought for the future is to not use the RPi to read it and use an Arduino connected to the RPi through some interface.

**Comment by kizniche** [comment]
I have several Pis reading ds18b20 in production systems that have run without issue for years. There's likely something about the setup of your systems that cause them to fail.

**Comment by mfittipaldi90** [comment]
I received the new Raspberry Pi and the sensors worked right away. It was my raspberry pi all along. Not sure it can be fixed…

We can close this ticket as resolved

**Comment by mfittipaldi90** [comment]
After a few days of without any issues, I had this behavior overnight. For 3 hours the temp sensor was gone and the EC/PH measurements were erratic/noisy. Even after the temperature sensor coming back online by itself, the measurements were fairly jumpy. At around 8:40 AM I performed a reboot of the raspberry pi and the measurements stabilized again. Not sure what could be causing this issue? Maybe a surge or drop in the wall outlet? I am running the 15 W Raspberry Pi power supply, and measure ~7 W usage with a power meter.

![image](https://github.com/user-attachments/assets/27b9e4a9-c63c-4a41-bb82-b430d83e3768)

**Comment by mfittipaldi90** [comment]
The power to the sensor is coming from the 3.3v or 5v pin of the raspberry pi. Do you recommend using a separate power source for it?

**Comment by mfittipaldi90** [comment]
For the pH and EC sensors, I am using an analog signal isolator in between them and the raspberry pi. Not sure if there is anything like that for the 1-wire. I could try to get an isolated external power supply for them

**Comment by kizniche** [comment]
Your DS sensor is submerged?

**Comment by mfittipaldi90** [comment]
Yes it is. Same reservoir as the pH and EC

**Comment by kizniche** [comment]
Do any erratic measurements cease when you remove any sensor or combination of sensors from the fluid?

**Comment by mfittipaldi90** [comment]
I saw that when the pH and EC sensors were together in the same res without the galvanic isolator circuits. Once I added those modules in, it was smooth. The overall behavior is stable and smooth overall. Not sure if it was a surge or spike in the outlet that friend the 1-wire on the RPI and that temporary loss on the second. I did add a surge protector to the outlet just in case...
![image](https://github.com/user-attachments/assets/4bea6ca8-db6b-4118-a838-70e96fafba8b)
